File: packages/unity_bridge_contracts/lib/src/bridge_errors.dart
/// Typed bridge parse / validation failures.
sealed class BridgeError implements Exception {
  const BridgeError(this.message);
  final String message;

  @override
  String toString() => 'BridgeError: $message';
}

final class UnknownSchemaVersionError extends BridgeError {
  UnknownSchemaVersionError(int version)
    : super('Unsupported schemaVersion: $version');
}

final class UnknownMessageTypeError extends BridgeError {
  UnknownMessageTypeError(String type) : super('Unknown message type: $type');
}

final class MissingRequiredFieldError extends BridgeError {
  MissingRequiredFieldError(String field)
    : super('Missing required field: $field');
}

final class StaleRevisionError extends BridgeError {
  StaleRevisionError({required this.incoming, required this.authoritative})
    : super('Stale revision: incoming=$incoming authoritative=$authoritative');

  final int incoming;
  final int authoritative;
}

final class DuplicateMessageError extends BridgeError {
  DuplicateMessageError(String id) : super('Duplicate message/request id: $id');
}

File: unity/ArabSolitaireUnity/Assets/ArabSolitaire/Bridge/BridgeErrors.cs
using System;

namespace ArabSolitaire.Bridge
{
    public abstract class BridgeError : Exception
    {
        protected BridgeError(string message) : base(message)
        {
        }
    }

    public sealed class UnknownSchemaVersionError : BridgeError
    {
        public UnknownSchemaVersionError(int version)
            : base($"Unsupported schemaVersion: {version}")
        {
        }
    }

    public sealed class UnknownMessageTypeError : BridgeError
    {
        public UnknownMessageTypeError(string type)
            : base($"Unknown message type: {type}")
        {
        }
    }

    public sealed class MissingRequiredFieldError : BridgeError
    {
        public MissingRequiredFieldError(string field)
            : base($"Missing required field: {field}")
        {
        }
    }

    public sealed class StaleRevisionError : BridgeError
    {
        public StaleRevisionError(int received, int authoritative)
            : base($"Stale revision {received}; authoritative is {authoritative}")
        {
        }
    }

    public sealed class DuplicateMessageError : BridgeError
    {
        public DuplicateMessageError(string id)
            : base($"Duplicate message/request id: {id}")
        {
        }
    }
}
